File: core/src/androidTest/java/com/emarsys/core/util/ImageUtilsTest.java
package com.emarsys.core.util;

import android.content.Context;
import android.graphics.Bitmap;
import android.graphics.BitmapFactory;

import com.emarsys.core.device.DeviceInfo;
import com.emarsys.core.device.LanguageProvider;
import com.emarsys.core.provider.hardwareid.HardwareIdProvider;
import com.emarsys.core.provider.version.VersionProvider;
import com.emarsys.testUtil.InstrumentationRegistry;
import com.emarsys.testUtil.TimeoutUtils;

import org.junit.Before;
import org.junit.Rule;
import org.junit.Test;
import org.junit.rules.TestRule;

import java.io.File;

import static com.emarsys.testUtil.TestUrls.LARGE_IMAGE;
import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertNotNull;
import static org.junit.Assert.assertNull;
import static org.junit.Assert.assertTrue;
import static org.mockito.Mockito.mock;

public class ImageUtilsTest {
    private Context context;
    private DeviceInfo deviceInfo;

    @Rule
    public TestRule timeout = TimeoutUtils.getTimeoutRule();

    @Before
    public void setup() {
        context = InstrumentationRegistry.getTargetContext();
        deviceInfo = new DeviceInfo(context,
                mock(HardwareIdProvider.class),
                mock(VersionProvider.class),
                mock(LanguageProvider.class));
    }

    @Test
    public void testLoadBitmap_contextShouldNotBeNull() {
        assertNull(ImageUtils.loadBitmap(null, "url"));
    }

    @Test
    public void testLoadBitmap_imageUrlShouldNotBeNull() {

        assertNull(ImageUtils.loadBitmap(context, null));
    }

    @Test
    public void testLoadBitmap_imageUrlShouldNotBeInvalid() {
        assertNull(ImageUtils.loadBitmap(context, "invalidUrl"));
    }

    @Test
    public void testLoadBitmap_withRemoteUrl() {
        Bitmap bitmap = ImageUtils.loadBitmap(context, LARGE_IMAGE);
        assertNotNull(bitmap);
        assertEquals(bitmap.getWidth(), 2500);
        assertEquals(bitmap.getHeight(), 2505);
    }

    @Test
    public void testLoadBitmap_withRemoteUrl_CleansUpTempFile() {
        clearCache();
        assertEquals(0, context.getCacheDir().list().length);
        ImageUtils.loadBitmap(context, LARGE_IMAGE);
        assertEquals(0, context.getCacheDir().list().length);
    }

    @Test
    public void testLoadBitmap_withLocalFile_ShouldNotCleansUpLocalFile() {
        clearCache();
        String fileUrl = FileUtils.download(context, LARGE_IMAGE);
        File imageFile = new File(fileUrl);
        assertTrue(imageFile.exists());
        ImageUtils.loadBitmap(context, fileUrl);
        assertTrue(imageFile.exists());
    }

    @Test(expected = IllegalArgumentException.class)
    public void testLoadOptimizedBitmap_deviceInfo_mustNotBeNull() {
        ImageUtils.loadOptimizedBitmap(context, LARGE_IMAGE, null);
    }

    @Test(expected = IllegalArgumentException.class)
    public void testLoadOptimizedBitmap_context_mustNotBeNull() {
        assertNull(ImageUtils.loadOptimizedBitmap(null, "url", deviceInfo));
    }

    @Test
    public void testLoadOptimizedBitmap_returnsNull_whenImageUrlIsNull() {
        assertNull(ImageUtils.loadOptimizedBitmap(context, null, deviceInfo));
    }

    @Test
    public void testLoadOptimizedBitmap_withRemoteUrl_CleansUpTempFile() {
        clearCache();
        assertEquals(0, context.getCacheDir().list().length);
        ImageUtils.loadOptimizedBitmap(context, LARGE_IMAGE, deviceInfo);
        assertEquals(0, context.getCacheDir().list().length);
    }

    @Test
    public void testLoadOptimizedBitmap_withLocalFile_ShouldNotCleanUpLocalFile() {
        clearCache();
        String fileUrl = FileUtils.download(context, LARGE_IMAGE);
        File imageFile = new File(fileUrl);
        assertTrue(imageFile.exists());
        ImageUtils.loadOptimizedBitmap(context, fileUrl, deviceInfo);
        assertTrue(imageFile.exists());
    }

    @Test
    public void testLoadOptimizedBitmap_imageUrlShouldBeInvalid() {
        assertNull(ImageUtils.loadOptimizedBitmap(context, "invalidUrl", deviceInfo));
    }

    @Test
    public void testLoadOptimizedBitmap_withRemoteUrl() {
        Bitmap bitmap = ImageUtils.loadOptimizedBitmap(context, LARGE_IMAGE, deviceInfo);
        assertNotNull(bitmap);
        assertTrue(bitmap.getWidth() < 2500);
        assertTrue(bitmap.getHeight() < 2505);
    }

    @Test
    public void testCalculateInSampleSize_returnedValueShouldBe4_whenRequestedWidthIs1080_widthIs2500() {
        BitmapFactory.Options options = new BitmapFactory.Options();
        options.outWidth = 2500;
        int inSampleSize = ImageUtils.calculateInSampleSize(options, 1080);
        assertEquals(4, inSampleSize);
    }

    private void clearCache() {
        clearDir(context.getCacheDir());
    }

    private void clearDir(File cacheDir) {
        for (File file : cacheDir.listFiles()) {
            if (file.isDirectory()) {
                clearDir(file);
            }
            file.delete();
        }
    }
}

File: core/src/androidTest/java/com/emarsys/core/util/ImageUtilsTest.kt
package com.emarsys.core.util

import android.graphics.BitmapFactory
import com.emarsys.core.device.DeviceInfo
import com.emarsys.core.device.LanguageProvider
import com.emarsys.core.notification.NotificationManagerHelper
import com.emarsys.core.provider.hardwareid.HardwareIdProvider
import com.emarsys.core.provider.version.VersionProvider
import com.emarsys.testUtil.InstrumentationRegistry.Companion.getTargetContext
import com.emarsys.testUtil.RetryUtils
import com.emarsys.testUtil.TimeoutUtils
import com.emarsys.testUtil.copyInputStreamToFile
import io.kotlintest.matchers.numerics.shouldBeLessThan
import io.kotlintest.shouldBe
import io.kotlintest.shouldNotBe
import org.junit.Before
import org.junit.Rule
import org.junit.Test
import org.junit.rules.TestRule
import org.mockito.Mockito
import org.mockito.kotlin.any
import org.mockito.kotlin.doAnswer
import org.mockito.kotlin.doReturn
import org.mockito.kotlin.mock
import java.io.File

class ImageUtilsTest {

    private companion object {
        const val IMAGE_URL = "https://emarsys.com"
    }

    private lateinit var mockFileDownloader: FileDownloader
    private lateinit var deviceInfo: DeviceInfo
    private lateinit var mockHardwareIdProvider: HardwareIdProvider
    private lateinit var mockLanguageProvider: LanguageProvider
    private lateinit var mockVersionProvider: VersionProvider

    @Rule
    @JvmField
    val timeout: TestRule = TimeoutUtils.timeoutRule

    @Rule
    @JvmField
    var retry: TestRule = RetryUtils.retryRule

    @Before
    fun setup() {
        mockFileDownloader = mock {
            on { download(any()) } doAnswer {
                val fileContent = getTargetContext().resources.openRawResource(
                        getTargetContext().resources.getIdentifier("emarsys_test_image",
                                "raw", getTargetContext().packageName))
                val file = File(getTargetContext().cacheDir.toURI().toURL().path + "/testFile.tmp")
                file.copyInputStreamToFile(fileContent)
                file.toURI().toURL().path
            }
            on { delete(any()) } doAnswer {
                File(getTargetContext().cacheDir.toURI().toURL().path + "/testFile.tmp").delete()
                Unit
            }

        }
        mockHardwareIdProvider = mock {
            on { provideHardwareId() } doReturn "hardwareId"
        }
        mockLanguageProvider = mock {
            on { provideLanguage(any()) } doReturn "language"
        }
        mockVersionProvider = mock {
            on { provideSdkVersion() } doReturn "version"
        }

        deviceInfo = DeviceInfo(
                getTargetContext(),
                mockHardwareIdProvider,
                mockVersionProvider,
                mockLanguageProvider,
                Mockito.mock(NotificationManagerHelper::class.java),
                isAutomaticPushSendingEnabled = true,
                isGooglePlayAvailable = true
        )
    }


    @Test
    fun testLoadOptimizedBitmap_returnsNull_whenImageUrlIsNull() {
        ImageUtils.loadOptimizedBitmap(mockFileDownloader, null, deviceInfo) shouldBe null
    }

    @Test
    fun testLoadOptimizedBitmap_withRemoteUrl_CleansUpTempFile() {
        clearCache()
        getTargetContext().cacheDir.list()?.size shouldBe 0
        ImageUtils.loadOptimizedBitmap(mockFileDownloader, IMAGE_URL, deviceInfo)
        getTargetContext().cacheDir.list()?.size shouldBe 0
    }

    @Test
    fun testLoadOptimizedBitmap_withLocalFile_ShouldNotCleanUpLocalFile() {
        clearCache()
        val fileUrl = mockFileDownloader.download(IMAGE_URL)
        val imageFile = File(fileUrl!!)
        imageFile.exists() shouldBe true
        ImageUtils.loadOptimizedBitmap(mockFileDownloader, fileUrl, deviceInfo)
        imageFile.exists() shouldBe true
    }

    @Test
    fun testLoadOptimizedBitmap_imageUrlShouldBeInvalid() {
        ImageUtils.loadOptimizedBitmap(mockFileDownloader, "invalidUrl", deviceInfo) shouldBe null
    }

    @Test
    fun testLoadOptimizedBitmap_withRemoteUrl() {
        val bitmap = ImageUtils.loadOptimizedBitmap(mockFileDownloader, IMAGE_URL, deviceInfo)
        bitmap shouldNotBe null
        bitmap!!.width shouldBeLessThan 2500
        bitmap.height shouldBeLessThan 2505
    }

    @Test
    fun testCalculateInSampleSize_returnedValueShouldBe4_whenRequestedWidthIs1080_widthIs2500() {
        val options = BitmapFactory.Options().apply {
            outWidth = 2500
        }
        val inSampleSize = ImageUtils.calculateInSampleSize(options, 1080)
        inSampleSize shouldBe 4
    }

    private fun clearCache() {
        clearDir(getTargetContext().cacheDir)
    }

    private fun clearDir(cacheDir: File) {
        for (file in cacheDir.listFiles() ?: emptyArray()) {
            if (file.isDirectory) {
                clearDir(file)
            }
            file.delete()
        }
    }
}